source/server/active_internal_listener.cc

```diff
@@ -53,8 +54,14 @@ void ActiveInternalListener::onAccept(Network::ConnectionSocketPtr&& socket) {
   // connections.
   incNumConnections();
 
+  auto* io_handle = dynamic_cast<Extensions::IoSocket::UserSpace::IoHandle*>(&socket->ioHandle());
```

Comment:
We can modify the interface ActiveInternalListener::onAccept(Network::ConnectionSocketPtr&& socket) to
 ActiveInternalListener::onAccept(Network::ConnectionSocketPtr&& socket, UserSpace::IoHandle) to avoid dynamic cast, 
but it doesn't resolve the polluting the deps because IoHandle is in extension.

I am wondering if we can list ActiveInternalListener as an exception for now

Reply:
I don't like this dependency loop. `ActiveInternalListener` is in the core but it only works with an extension `io_socket/user_space`. That shouldn't be allowed. We need to either move both to extension or make `io_socket/user_space` be effectively in core. cc @alyssawilk

Reply:
I'd be inclined to make both extensions, especially if Envoy can build without the active internal listener. IMO it ideally would be optional and something folks who don't need can leave out of the binary.

Reply:
This is a non-trivial cleanup, filed https://github.com/envoyproxy/envoy/issues/20798.

Reply:
I'll address the dependency loop separately. It requires some core changes that shouldn't be coupled with this PR.